string.go

```diff
@@ -20,18 +18,46 @@ type StringRule struct {
 func NewStringRule(validator stringValidator, message string) StringRule {
 	return StringRule{
 		validate: validator,
-		message:  message,
+		err:      NewError("", message),
+	}
+}
+
+// NewStringRuleWithError creates a new validation rule using a function that takes a string value and returns a bool.
+// The rule returned will use the function to check if a given string or byte slice is valid or not.
+// An empty value is considered to be valid. Please use the Required rule to make sure a value is not empty.
+func NewStringRuleWithError(validator stringValidator, err Error) StringRule {
+	return StringRule{
+		validate: validator,
+		err:      err,
 	}
 }
 
 // Error sets the error message for the rule.
-func (v StringRule) Error(message string) StringRule {
-	v.message = message
-	return v
+func (r StringRule) Error(message string) StringRule {
+	r.err.SetMessage(message)
+	return r
+}
+
+// ErrorObject sets the error struct for the rule.
+func (r StringRule) ErrorObject(err Error) StringRule {
+	r.err = err
+	return r
+}
+
+// ErrParams sets the rule's error params.
+func (r StringRule) ErrParams(params map[string]interface{}) StringRule {
+	r.err.SetParams(params)
+	return r
+}
+
+// Code sets the rule's translation code (translation key).
+func (r StringRule) Code(code string) StringRule {
+	r.err.SetCode(code)
+	return r
 }
```

Comment:
I think we should still remove the above two methods `ErrParams` and `Code`. If a user wants to set them (which is very uncommon), he can still achieve it via `ErrorObject()`.

Reply:
Yeah they can achieve it via `ErrorObject()`, but their code change from e.g :

```go
validation.Email.Code("v_invalid_email")
```
to 
```go
errEmail:=validation.ErrEmail
errEmail.SetCode("v_invalid_email")
validation.Email.ErrorObject(errEmail)
```

Are you ok with it?

Reply:
I can hardly think of any reason that a user would want to change error code.

Reply:
OK, so I remove them.